core/src/sumeragi/network_topology.rs

```diff
@@ -5,341 +5,183 @@
-    pub fn proxy_tail(&self) -> &PeerId {
-        self.peers_set_a().last().expect("Failed to get last peer.")
-    }
-
-    /// The peers that validate the block in discussion this round and vote for it to be accepted by the blockchain.
-    pub fn validating_peers(&self) -> &[PeerId] {
-        let a_set = self.peers_set_a();
-        if a_set.len() > 1 {
-            &a_set[1..(a_set.len() - 1)]
-        } else {
-            &[]
-        }
+    /// How many faulty peers can this topology tolerate.
+    pub fn max_faults(&self) -> usize {
+        (self.sorted_peers.len().saturating_sub(1)) / 3
     }
-
-    /// Get role of the peer by its id.
-    pub fn role(&self, peer_id: &PeerId) -> Role {
-        if self.leader() == peer_id {
-            Role::Leader
-        } else if self.proxy_tail() == peer_id {
-            Role::ProxyTail
-        } else if self.validating_peers().contains(peer_id) {
-            Role::ValidatingPeer
-        } else {
-            Role::ObservingPeer
-        }
+    /// The required amount of votes to commit a block with this topology.
+    pub fn min_votes_for_commit(&self) -> usize {
+        self.max_faults() * 2 + 1
     }
-
-    /// Verifies that this `message` was signed by the `signature` of a peer with specified `role`.
-    ///
-    /// # Errors
-    /// Fails if there are no such peer with this key and if signature verification fails
-    // TODO: Not used?
-    pub fn verify_signature_with_role(
+    /// Filter signatures by roles in the topology.
+    #[allow(clippy::comparison_chain)]
+    pub fn filter_signatures_by_roles<'a, T: 'a, I: IntoIterator<Item = &'a SignatureOf<T>>>(
         &self,
-        signature: &SignatureOf<VersionedSignedTransaction>,
-        role: Role,
-        tx: &HashOf<VersionedSignedTransaction>,
-    ) -> Result<()> {
-        if role
-            .peers(self)
-            .iter()
-            .any(|peer| peer.public_key == *signature.public_key())
-        {
-            Ok(())
-        } else {
-            Err(eyre!("No {:?} with this public key exists.", role))
-        }
-        .and_then(|()| {
-            signature
-                .verify_hash(tx)
-                .wrap_err("Transaction signature check failed")
-        })
-    }
-
-    /// Returns signatures of the peers with the specified `roles` from all `signatures`.
-    pub fn filter_signatures_by_roles<'slf, T: 'slf>(
-        &'slf self,
-        roles: &'slf [Role],
-        signatures: impl IntoIterator<Item = &'slf SignatureOf<T>> + 'slf,
+        roles: &[Role],
+        signatures: I,
     ) -> Vec<SignatureOf<T>> {
-        let roles: HashSet<Role> = roles.iter().copied().collect();
-        let public_keys: HashSet<_> = roles
-            .iter()
-            .flat_map(|role| role.peers(self))
-            .map(|peer| peer.public_key)
-            .collect();
+        let mut public_keys = Vec::new();
+        for role in roles {
+            let role_public_keys = if self.max_faults() == 0 {
```

Comment:
You can match on a tuple. e.g. 

```rust 
match (role, self.max_faults()) { 
   (Role::Leader, 0) => vec![self.sorted_peers[0].public_key.clone()],
   // which you can only see because of matching twice. 
   (Role::Leader, _) => vec![self.sorted_peers[0].public_key.clone()]
   // And similarly for other patterns. 
}

```

Reply:
sure